documentation/writing_tests.md

```diff
@@ -0,0 +1,1534 @@
+
+- The titles sent to the `describe` calls should represent what the collection of its inner test cases (`it` calls) are testing.
+- The titles sent to each `it` call should represent what each specific test case is testing.
+
+Therefore, `describe` titles should be written to reflect something that each of its `it` test cases (or inner `describe` blocks) share in common.
+
+Let's build up an example. Given the following code, which will exist in a file named `internalClass.ts`:
+
+```ts
+export class InternalClass {
+  returnsTrue(): boolean {
+    return true;
+  }
+}
+
+export class InternalInheritedClass extends InternalClass {
+  returnsFalse(): boolean {
+    return false;
+  }
+}
+```
+
+We could group all of the possible tests for that pair of classes in a single file. Let's say we call this test file `internalClass.spec.ts`. The outermost `describe` would have a general description of what are these classes related to. If these are all of our internal classes, we can set the title as `Tests for the internal classes`. Inside of that `describe`, we can set two inner `describe` blocks, one for `InternalClass` and another one for `InternalInheritedClass`, as follows:
+
+```ts
+describe("Tests for the internal classes", function() {
+
+  describe("Tests for the InternalClass", function() {
+    // ...
+  });
+
+  describe("Tests for the InternalInheritedClass", function() {
+    // ...
+  });
+
+});
+```
+
+Inside of the inner describes, each test case would focus on asserting one specific behavior. For `InternalClass`, we would first assert that the class' constructor works as expected, then test it's `returnsTrue` method. For `InternalInheritedClass`, we would first assert that the class' constructor works as expected, then test it's inherited `returnsTrue` method, then it's `returnFalse` method. As follows:
+
+```ts
+describe("Tests for the internal classes", function() {
+
+  describe("Tests for the InternalClass", function() {
+    it("The InternalClass should be able to be initialized", function() {
+      const Internal = new InternalClass();
+      assert.exists(Internal);
+    });
+
+    it("The InternalClass's returnsTrue should return true", function() {
+      const Internal = new InternalClass();
+      const result: boolean = Internal.returnsTrue();
+      assert.isTrue(result);
+    });
+  });
+
+  describe("Tests for the InternalInheritedClass", function() {
+    it("The InternalInheritedClass should be able to be initialized", function() {
+      const Internal = new InternalInheritedClass();
+      assert.exists(Internal);
+    });
+
+    it("The InternalInheritedClass's returnsTrue should return true", function() {
+      const Internal = new InternalInheritedClass();
+      const result: boolean = Internal.returnsTrue();
```

Comment:
in general do we recommend testing non-shadowing inherited members? seems unnecessary

Reply:
I'm mostly concerned about displaying the functionality to our users. Is it enough to let them figure out or would it help being verbose about what each class is capable of doing?

Reply:
I guess my instinct is you don't bother testing inherited members unless you do something that uniquely changes the way they act. Otherwise you're just testing that inheritance works, which hopefully isn't something we need to do.

Reply:
I think what you're saying needs to be explicitly mentioned in the guidelines! I'll make sure it is, and I will make sure the tests reflect it 🌞

Reply:
I think I addressed this now! I'll resolve this. Let me know if I should change it!